### PR TITLE
Profiling

### PR DESCRIPTION
**Pull request by Kishore-B-Rao**
Took over [WIP] Profiling of phases #650 and fixed some of the issues. Rebased to the most recent code and added more profiling information.

### CODE REVIEW

File: src/serializer/serializer.js [hunk 1]
```diff
@@ -1842,34 +1841,43 @@ export class Serializer {
   init(filename: string, code: string, map?: string = "",
       sourceMaps?: boolean = false, onError?: (Realm, Value) => void) {
     // Phase 1: Let's interpret.
+    if (this.options.profile) console.time("Speculative Initialization");
     this.execute(filename, code, map, onError);
     if (this.logger.hasErrors()) return undefined;
     this.modules.resolveInitializedModules();
     if (this.options.initializeMoreModules) {
```
Comment (hermanventer) [inline src/serializer/serializer.js]: This is where you should log the start of speculative initialization.

File: src/serializer/serializer.js [hunk 2]
```diff
@@ -1842,34 +1841,43 @@ export class Serializer {
   init(filename: string, code: string, map?: string = "",
       sourceMaps?: boolean = false, onError?: (Realm, Value) => void) {
     // Phase 1: Let's interpret.
+    if (this.options.profile) console.time("Speculative Initialization");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: This is the start of non speculative initialization.

File: src/serializer/serializer.js [hunk 3]
```diff
@@ -1842,34 +1841,43 @@ export class Serializer {
...
     this.execute(filename, code, map, onError);
     if (this.logger.hasErrors()) return undefined;
     this.modules.resolveInitializedModules();
     if (this.options.initializeMoreModules) {
       this.modules.initializeMoreModules();
       if (this.logger.hasErrors()) return undefined;
     }
+    if (this.options.profile) console.timeEnd("Speculative Initialization");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: Move this up a line.

File: src/serializer/serializer.js [hunk 4]
```diff
@@ -1842,34 +1841,43 @@ export class Serializer {
...
     this.ignoredProperties = residualHeapVisitor.ignoredProperties;
     this.residualValues = residualHeapVisitor.values;
     this.residualFunctionBindings = residualHeapVisitor.functionBindings;
     this.residualFunctionInfos = residualHeapVisitor.functionInfos;
+    if (this.options.profile) console.timeEnd("Deep Traversal of Heap");
 
     // Phase 2: Let's serialize the heap and generate code.
     // Serialize for the first time in order to gather reference counts
     if (!this.options.singlePass) {
+      if (this.options.profile) console.time("Optional First Serialize Pass");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: Rather call this the reference counts pass.

File: src/serializer/serializer.js [hunk 5]
```diff
@@ -1842,34 +1841,43 @@ export class Serializer {
...
       this.collectValToRefCountOnly = true;
       this.valToRefCount = new Map();
       this.serialize(filename, code, sourceMaps);
       if (this.logger.hasErrors()) return undefined;
+      if (this.options.profile) console.timeEnd("Optional First Serialize Pass");
     }
     // Serialize for a second time, using reference counts to minimize number of generated identifiers
     this._resetSerializeStates();
     this.collectValToRefCountOnly = false;
+    if (this.options.profile) console.time("Second Serialize Pass");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: And call this just the serialize pass.

File: src/serializer/serializer.js [hunk 6]
```diff
@@ -1842,34 +1841,43 @@ export class Serializer {
   init(filename: string, code: string, map?: string = "",
       sourceMaps?: boolean = false, onError?: (Realm, Value) => void) {
     // Phase 1: Let's interpret.
+    if (this.options.profile) console.time("Speculative Initialization");
     this.execute(filename, code, map, onError);
     if (this.logger.hasErrors()) return undefined;
     this.modules.resolveInitializedModules();
```
Comment (NTillmann) [inline src/serializer/serializer.js]: The call to `resolveInitializedModules` is potentially expensive, and should get its own time tracking.

File: src/serializer/serializer.js [hunk 7]
```diff
@@ -1842,34 +1841,43 @@ export class Serializer {
   init(filename: string, code: string, map?: string = "",
       sourceMaps?: boolean = false, onError?: (Realm, Value) => void) {
     // Phase 1: Let's interpret.
+    if (this.options.profile) console.time("Speculative Initialization");
     this.execute(filename, code, map, onError);
```
Comment (NTillmann) [inline src/serializer/serializer.js]: Just `execute` should be time tracked as "Interpreting global code".

File: src/serializer/serializer.js [hunk 8]
```diff
@@ -1842,34 +1841,43 @@ export class Serializer {
...
     if (this.logger.hasErrors()) return undefined;
     this.modules.resolveInitializedModules();
     if (this.options.initializeMoreModules) {
       this.modules.initializeMoreModules();
```
Comment (NTillmann) [inline src/serializer/serializer.js]: `initializeMoreModules` is the essence of "speculative initialization".

**Comment by Kishore-B-Rao** [comment]
The profiling summary can be extracted from all the other output by piping output into grep "\[Profiling\]"

File: src/serializer/serializer.js [hunk 9]
```diff
@@ -1842,34 +1841,47 @@ export class Serializer {
   init(filename: string, code: string, map?: string = "",
       sourceMaps?: boolean = false, onError?: (Realm, Value) => void) {
     // Phase 1: Let's interpret.
+    if (this.options.profile) console.time("[Profiling] Interpreting Global Code");
     this.execute(filename, code, map, onError);
```
Comment (NTillmann) [inline src/serializer/serializer.js]: Would be nice if you could extract the "Babel parsing" part of execute. But that can come in a separate pull request.

File: src/serializer/serializer.js [hunk 10]
```diff
@@ -1842,34 +1841,47 @@ export class Serializer {
...
       this.valToRefCount = new Map();
       this.serialize(filename, code, sourceMaps);
       if (this.logger.hasErrors()) return undefined;
+      if (this.options.profile) console.timeEnd("[Profiling] Reference Counts Pass");
     }
     // Serialize for a second time, using reference counts to minimize number of generated identifiers
     this._resetSerializeStates();
     this.collectValToRefCountOnly = false;
+    if (this.options.profile) console.time("[Profiling] Serialize Pass");
     let serialized = this.serialize(filename, code, sourceMaps);
```
Comment (NTillmann) [inline src/serializer/serializer.js]: Would be nice if you could extract the "Babel generating" part of serialize. But that can come in a separate pull request.

File: src/prepack-cli.js [hunk 11]
```diff
@@ -35,6 +35,7 @@ function run(Object, Array, console, JSON, process, prepackStdin, prepackFileSyn
     --trace    Traces the order of module initialization.
     --serialize    Serializes the partially evaluated global environment as a program that recreates it. (default = true)
     --residual    Produces the residual program that results after constant folding.
+    --profile    Enables profile console logging of Interpret phase, first and second Serialize phase.
```
Comment (NTillmann) [inline src/prepack-cli.js]: Sounds awkward. Please rephrase.

**Pull request merged**